x-pack/plugins/endpoint/server/routes/resolver/utils/fetch.ts

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { IScopedClusterClient } from 'kibana/server';
+import { extractEntityID, extractParentEntityID } from './normalize';
+import { getPaginationParams } from './pagination';
+import { Tree } from './tree';
+import { LifecycleQuery } from '../queries/lifecycle';
+import { ChildrenQuery } from '../queries/children';
+import { EventsQuery } from '../queries/events';
+import { StatsQuery } from '../queries/stats';
+
+export class Fetcher {
+  constructor(
+    private readonly client: IScopedClusterClient,
+    private readonly id: string,
+    private readonly endpointID?: string
+  ) {}
+
+  public async ancestors(limit: number): Promise<Tree> {
+    const tree = new Tree(this.id);
+    await this.doAncestors(tree, this.id, this.id, limit);
+    return tree;
+  }
+
+  public async children(limit: number, generations: number, after?: string): Promise<Tree> {
+    const tree = new Tree(this.id);
+    await this.doChildren(tree, [this.id], limit, generations, after);
+    return tree;
+  }
+
+  public async events(limit: number, after?: string): Promise<Tree> {
```

Comment:
This the events returned by the query are all associated with a single node and not really making use of the tree structure.  It would be easier to follow if this returned a list of events which was then added to the children tree by id.

Reply:
We shouldn't be querying for events associated with multiple processes at a time due to:

1. Only needing to show events when a node is in focus
2. The fact that a single process can spawn thousands if not more events.

The tree structure is there to make marking pagination parameters possible on a per-node basis, and to facilitate client-side merging of all of the return values of these trees.

Reply:
Sure, I'm not at all suggesting that we allow querying for events from multiple processes simultaneously.  I added a comment in my review about how we should prevent querying for events from multiple processes.

My point is that the result of this function is a list of events which all share an entity_id so returning it as a tree structure with unused parent and children structures is unnecessary and somewhat confusing.  My first thought reading it is "i expect this to be a list of siblings, so what are the parents/children being used for?"